![Publish docs via GitHub Pages](https://github.com/birneylab/MIKK_genome_companion_paper/workflows/Publish%20docs%20via%20GitHub%20Pages/badge.svg)


---

# Welcome to the MIKK panel genome companion paper repository

---

**Repository for the forthcoming paper describing the Medaka Inbred Kiyosu-Karlsruhe (MIKK) panel genome companion paper by the Birney Group (EMBL-EBI), Wittbrodt Group (COS Heidelberg), and Loosli Group (KIT).**

### Study Abstract

Oryzias latipes (medaka) has the highly attractive property that it can be routinely inbred to near isogenic genomes from wild founder individuals. An inbred panel resource known as the Medaka Inbred Kiyosu-Karlsruhe (MIKK) panel has been created exploiting this feature. Medaka is also a  well established model system having a long history of genetic research with multiple high quality reference genomes available from medaka populations (HdrR, HNI and HSOK). Here we used nanopore long read data to investigate further the genomic and epigenomic landscapes of a subset of the MIKK panel. Nanopore sequencing allowed us to identify a much greater variety of high quality structural variants compared with Illumina sequencing. We also present results and methods using a pan-genome graph representation of 12 individual medaka lines from the MIKK panel. By creating a graph based reference genome for the MIKK panel we are able to find novel differences between the MIKK panel lines compared to standard linear reference genomes and represent additional MIKK panel specific genomic content that would be missing from linear reference alignment approaches. We were also able to identify and quantify the presence of repeat elements such as the Teratorn fusion transposon, in each of the lines. Finally, we investigated line specific CpG methylation and performed differential DNA methylation analysis across the 12 lines analysis different lines. Overall we present a detailed analysis of the MIKK panel genomes using long and short read sequence technologies, creating a MIKK panel specific pan genome reference dataset allowing the investigation of novel variation types that would be elusive using standard approaches.

### Content

This repository contains the code and final results for the MIKK panel genome analyses including:

* [Nanopore basecalling](https://birneylab.github.io/MIKK_genome_companion_paper/Nanopore_basecalling)
* [Nanopore data processing](https://birneylab.github.io/MIKK_genome_companion_paper/Nanopore_processing)
* [Structural variations Nanopore](https://birneylab.github.io/MIKK_genome_companion_paper/Nanopore_SV_analysis)
* [DNA methylation analysis](https://birneylab.github.io/MIKK_genome_companion_paper/DNA_methylation)
* [Individual MIKK lines hybrid assemblies](https://birneylab.github.io/MIKK_genome_companion_paper/Individual_assemblies)
* [Pangenome graph assembly](https://birneylab.github.io/MIKK_genome_companion_paper/Graph_assembly)
* [Pangenome graph usage analysis](https://birneylab.github.io/MIKK_genome_companion_paper/Graph_usage)
* [Pangenome graph SV analysis](https://birneylab.github.io/MIKK_genome_companion_paper/Graph_SV)

### Citation

**Genomic variations and epigenomic landscape of the MIKK panel**, Adrien Leger#, Ian Brettell#, Jack Monahan, Bettina Welz, Carl Barton, Cathrin Herder, Clara Becker, Colin Lischik, Erika Tsingos, Eva Hasel, Jakob Gierten, Katharina Lust, Nadia Khoudja, Narendar Aadepu, Omar T. Hammouda, Philip Watson, Thomas Thumberger, Tinatini Tavhelidse, Kiyoshi Naruse, Felix Loosli, Joachim Wittbrodt, Ewan Birney, Tomas Fitzgeraldâ€ 
